compiler-core/src/language_server/completer.rs

```diff
@@ -916,3 +967,120 @@ fn field_completion(label: &str, type_: Arc<Type>) -> CompletionItem {
         ..Default::default()
     }
 }
+
+pub struct LocalCompletion<'a> {
+    mod_name: &'a str,
+    insert_range: Range,
+    cursor: u32,
+    completions: Vec<CompletionItem>,
+}
+
+impl<'a> LocalCompletion<'a> {
+    pub fn new(mod_name: &'a str, insert_range: Range, cursor: u32) -> Self {
+        Self {
+            mod_name,
+            insert_range,
+            cursor,
+            completions: Vec::new(),
+        }
+    }
+
+    pub fn completions(&self) -> Vec<CompletionItem> {
+        self.completions.clone()
+    }
+
+    pub fn visit_statements(&mut self, statements: &'a [Statement<Arc<Type>, TypedExpr>]) {
+        for statement in statements {
+            if statement.location().start >= self.cursor {
+                continue;
+            }
+            ast::visit::visit_typed_statement(self, statement);
+        }
+    }
+
+    pub fn visit_fn_args(&mut self, args: &[Arg<Arc<Type>>]) {
+        for arg in args {
+            if let Some(name) = arg.get_variable_name() {
+                self.push_completion(name, arg.type_.clone());
+            }
+        }
+    }
+
+    fn push_completion(&mut self, name: &str, type_: Arc<Type>) {
+        if name.is_empty() || name.starts_with('_') {
+            return;
+        }
+
+        self.completions.push(local_value_completion(
+            self.mod_name,
+            name,
+            type_,
+            self.insert_range,
+        ));
+    }
+}
+
+impl<'ast> ast::visit::Visit<'ast> for LocalCompletion<'_> {
+    fn visit_typed_assignment(&mut self, assignment: &'ast ast::TypedAssignment) {
```

Comment:
Could you use detailed comments to document why this function does this please 🙏  I didn't understand it at first.

Reply:
Added documentation! Let me know what you think if it is clear or not

Reply:
Wonderful!